src/main/java/htsjdk/variant/variantcontext/VariantContext.java

```diff
@@ -1638,106 +1651,221 @@ public boolean isFullyDecoded() {
-        builder.attributes(fullyDecodeAttributes(getAttributes(), header, lenientDecoding));
+    private void decodeInfo(
+        final VariantContextBuilder builder,
+        final VCFHeader header,
+        final boolean lenientDecoding
+    ) {
+        decodeAttributes(builder::attribute, id -> getInfoLine(id, header, lenientDecoding), this.getAttributes(), lenientDecoding);
     }
 
-    private final Map<String, Object> fullyDecodeAttributes(final Map<String, Object> attributes,
-                                                            final VCFHeader header,
-                                                            final boolean lenientDecoding) {
-        final Map<String, Object> newAttributes = new HashMap<>(10);
-
-        for ( final Map.Entry<String, Object> attr : attributes.entrySet() ) {
-            final String field = attr.getKey();
-
-            if ( field.equals(VCFConstants.GENOTYPE_FILTER_KEY) )
-                continue; // gross, FT is part of the extended attributes
-
-            final VCFCompoundHeaderLine format = VariantContextUtils.getMetaDataForField(header, field);
-            final Object decoded = decodeValue(field, attr.getValue(), format);
-
-            if ( decoded != null &&
-                    ! lenientDecoding
-                    && format.getCountType() != VCFHeaderLineCount.UNBOUNDED
-                    && format.getType() != VCFHeaderLineType.Flag ) { // we expect exactly the right number of elements
-                final int obsSize = decoded instanceof List ? ((List) decoded).size() : 1;
-                final int expSize = format.getCount(this);
-                if ( obsSize != expSize ) {
-                    throw new TribbleException.InvalidHeader("Discordant field size detected for field " +
-                            field + " at " + getContig() + ":" + getStart() + ".  Field had " + obsSize + " values " +
-                            "but the header says this should have " + expSize + " values based on header record " +
-                            format);
+    private void decodeGenotypes(
+        final VariantContextBuilder builder,
+        final VCFHeader header,
+        final boolean lenientDecoding
+    ) {
+        final GenotypesContext gc = this.getGenotypes();
+        final GenotypesContext newGc = new GenotypesContext(gc.size());
+        for (final Genotype g : this.getGenotypes()) {
+            final GenotypeBuilder gb = new GenotypeBuilder(g);
+            decodeAttributes(gb::attribute, id -> getFormatLine(id, header, lenientDecoding), g.getExtendedAttributes(), lenientDecoding);
+            newGc.add(gb.make());
+        }
+        builder.genotypesNoValidation(newGc);
+    }
+
+    private void decodeAttributes(
+        final BiConsumer<String, Object> put,
+        final Function<String, VCFCompoundHeaderLine> metadataGetter,
+        final Map<String, Object> attributes,
+        final boolean lenientDecoding
```

Comment:
We should try find some way to make this code type safe. Its not obvious from the argument types that the `put` and `attributes` args are related, and need to be type-compatible, and its a bit awkward to require the caller to deconstruct the source objects that own the maps and pass the pieces in pairs like this.

One suggestion to consider would be delegating this to new `getDecodedAttributes` methods on `Genotype` and `VariantContext`, since those classes own the map/put pairs that are passed in here together. Or maybe there is some other alternative.

But if thats awkward, I would at least reorder the args so that `attributes` is first, followed by the related `put`, and document how they're related.